File: app/lannsd/src/main/java/org/exthmui/share/lannsd/NsdReceivingTask.java
package org.exthmui.share.lannsd;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.exthmui.share.shared.base.Entity;
import org.exthmui.share.shared.base.FileInfo;
import org.exthmui.share.shared.base.IConnectionType;
import org.exthmui.share.shared.base.receive.ReceivingTask;
import org.exthmui.share.shared.base.receive.SenderInfo;
import org.exthmui.share.shared.base.results.TransmissionResult;
import org.exthmui.share.shared.exceptions.trans.InvalidInputDataException;
import org.exthmui.share.shared.exceptions.trans.TransmissionException;
import org.exthmui.share.shared.listeners.OnReceiveActionAcceptListener;
import org.exthmui.share.shared.listeners.OnReceiveActionRejectListener;
import org.exthmui.share.shared.misc.Constants;
import org.exthmui.share.shared.misc.ReceiverUtils;
import org.exthmui.share.taskMgr.Result;
import org.exthmui.share.udptransport.UDPReceiver;

import java.net.ServerSocket;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicReference;

public class NsdReceivingTask extends ReceivingTask {

    public static final String TAG = "NsdReceivingTask";

    public NsdReceivingTask(Context context, Bundle inputData) {
        super(context, inputData);
    }

    @NonNull
    @Override
    public IConnectionType getConnectionType() {
        return new Metadata();
    }

    @Nullable
    private final ServerSocket serverSocketToServer = null;

    @NonNull
    @Override
    public Result doWork() {
        int timeout = NsdUtils.getTimeout(getApplicationContext());

        final AtomicReference<Boolean> cancelledBySender = new AtomicReference<>(false);
        final AtomicReference<Boolean> completed = new AtomicReference<>(false);
        final AtomicReference<TransmissionResult> generalResult = new AtomicReference<>(null);
        final AtomicReference<Map<String, TransmissionResult>> resultMap = new AtomicReference<>(null);

        Bundle input = getInputData();
        byte connId = input.getByte(org.exthmui.share.lannsd.Constants.WORKER_INPUT_KEY_CONN_ID, (byte) 0);
        if (connId == 0)
            return genFailureResult(new InvalidInputDataException(getApplicationContext()), null);
        NsdReceiver receiver = NsdReceiver.getInstance(getApplicationContext());
        UDPReceiver udpReceiver = receiver.getUdpReceiver();
        if (udpReceiver == null) return genSilentResult();
        UDPReceiver.ConnectionHandler handler = udpReceiver.getHandler(connId);

        if (handler == null)
            return genFailureResult(new InvalidInputDataException(getApplicationContext()), null);

        handler.setListener(new UDPReceiver.ReceivingListener() {
            @Override
            public void requestAcceptationAsync(SenderInfo senderInfo, FileInfo[] fileInfos, CompletableFuture<Set<String>> idsAccepted) {
                // Wait for acceptation from user
                NsdReceiver.getInstance(getApplicationContext()).registerListener((OnReceiveActionAcceptListener) event -> {
                    Log.d(TAG, "User accepted file");
                    Set<String> ids = new HashSet<>();
                    for (FileInfo fileInfo : fileInfos) {
                        ids.add(fileInfo.getId());
                    }
                    idsAccepted.complete(ids);
                });
                NsdReceiver.getInstance(getApplicationContext()).registerListener((OnReceiveActionRejectListener) event -> {
                    Log.d(TAG, "User rejected file");
                    idsAccepted.complete(Collections.emptySet());
                });
                ReceiverUtils.requestAcceptation(getApplicationContext(),
                        Constants.CONNECTION_CODE_LANNSD, getTaskId(), senderInfo,
                        fileInfos, getTaskId().hashCode());
            }

            @Override
            public void onProgressUpdate(int status, long totalBytesToSend, long bytesReceived,
                                         @NonNull SenderInfo senderInfo,
                                         @NonNull FileInfo[] fileInfos, String curFileId,
                                         long curFileBytesToSend, long curFileBytesReceived) {
                updateProgress(status, totalBytesToSend, bytesReceived, fileInfos, senderInfo,
                        curFileId, curFileBytesToSend, curFileBytesReceived, totalBytesToSend == 0);
            }

            @Override
            public void onComplete(TransmissionResult result, Map<String, TransmissionResult> r) {
                generalResult.set(result);
                resultMap.set(r);
                completed.set(true);
            }
        });

        while (!completed.get()) {
            // Check if user cancelled
            if (isCancelled()) {
                Log.d(TAG, "User cancelled receiving file");
                handler.cancel();
            }
        }

        if (generalResult.get() == null || generalResult.get().getStatus() != Constants.TransmissionStatus.COMPLETED) {
            return genFailureResult((TransmissionException) generalResult.get(), handler.getSenderInfo(), handler.getFileInfos(),
                    resultMap.get());
        }

        NsdReceiver nsdReceiver = NsdReceiver.getInstance(getApplicationContext());

        List<Entity> entities = new ArrayList<>(handler.getFileInfos().length);
        for (FileInfo fileInfo : handler.getFileInfos()) {
            entities.add(nsdReceiver.getEntity(fileInfo.getId()));
        }
        return genSuccessResult(handler.getSenderInfo(), entities);
    }
}

File: app/lannsd/src/main/java/org/exthmui/share/lannsd/NsdReceivingTask.kt
package org.exthmui.share.lannsd

import android.content.Context
import android.os.Bundle
import android.util.Log
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.coroutineScope
import org.exthmui.share.shared.base.Entity
import org.exthmui.share.shared.base.FileInfo
import org.exthmui.share.shared.base.IConnectionType
import org.exthmui.share.shared.base.receive.ReceivingTask
import org.exthmui.share.shared.base.receive.SenderInfo
import org.exthmui.share.shared.base.results.TransmissionResult
import org.exthmui.share.shared.events.ReceiveActionAcceptEvent
import org.exthmui.share.shared.events.ReceiveActionRejectEvent
import org.exthmui.share.shared.exceptions.trans.InvalidInputDataException
import org.exthmui.share.shared.exceptions.trans.TransmissionException
import org.exthmui.share.shared.listeners.OnReceiveActionAcceptListener
import org.exthmui.share.shared.listeners.OnReceiveActionRejectListener
import org.exthmui.share.shared.misc.ReceiverUtils
import org.exthmui.share.taskMgr.Result
import org.exthmui.share.taskMgr.entities.TaskEntity
import org.exthmui.share.udptransport.UDPReceiver.ReceivingListener
import java.net.ServerSocket
import java.util.concurrent.CompletableFuture
import java.util.concurrent.atomic.AtomicReference

class NsdReceivingTask : ReceivingTask {

    constructor(context: Context?, inputData: Bundle?) : super(context!!, inputData)
    constructor(taskEntity: TaskEntity) : super(taskEntity)

    override val connectionType: IConnectionType
        get() = Metadata()

    private val serverSocketToServer: ServerSocket? = null

    override suspend fun doWork(): Result = coroutineScope {
        applicationContext?.let {
            val timeout = NsdUtils.getTimeout(it)
            val cancelledBySender = AtomicReference(false)
            val completed = CompletableDeferred<Boolean>()
            val generalResult = AtomicReference<TransmissionResult?>(null)
            val resultMap = AtomicReference<Map<String, TransmissionResult>?>(null)
            val input = inputData
            val connId = input.getByte(Constants.WORKER_INPUT_KEY_CONN_ID, 0.toByte())
            if (connId.toInt() == 0) return@coroutineScope genFailureResult(
                InvalidInputDataException(
                    it
                ), null
            )
            val receiver = NsdReceiver.getInstance(it)
            val udpReceiver = receiver.udpReceiver ?: return@coroutineScope genSilentResult()
            val handler = udpReceiver.getHandler(connId)
                ?: return@coroutineScope genFailureResult(InvalidInputDataException(it), null)
            handler.setListener(object : ReceivingListener {
                override fun requestAcceptationAsync(
                    senderInfo: SenderInfo,
                    fileInfos: Array<FileInfo>,
                    idsAccepted: CompletableFuture<Set<String>>?
                ) {
                    // Wait for acceptation from user
                    NsdReceiver.getInstance(it).registerListener(
                        OnReceiveActionAcceptListener { _: ReceiveActionAcceptEvent? ->
                            Log.d(TAG, "User accepted file")
                            val ids: MutableSet<String> = HashSet()
                            for (fileInfo in fileInfos) {
                                ids.add(fileInfo.id)
                            }
                            idsAccepted!!.complete(ids)
                        }
                    )
                    NsdReceiver.getInstance(it).registerListener(
                        OnReceiveActionRejectListener { event: ReceiveActionRejectEvent? ->
                            Log.d(TAG, "User rejected file")
                            idsAccepted!!.complete(emptySet())
                        }
                    )
                    ReceiverUtils.requestAcceptation(
                        it,
                        org.exthmui.share.shared.misc.Constants.CONNECTION_CODE_LANNSD,
                        taskId,
                        senderInfo,
                        fileInfos,
                        taskId.hashCode()
                    )
                }

                override fun onProgressUpdate(
                    status: Int, totalBytesToSend: Long, bytesReceived: Long,
                    senderInfo: SenderInfo,
                    fileInfos: Array<FileInfo>, curFileId: String?,
                    curFileBytesToSend: Long, curFileBytesReceived: Long
                ) {
                    updateProgress(
                        status,
                        totalBytesToSend,
                        bytesReceived,
                        fileInfos,
                        senderInfo,
                        curFileId,
                        curFileBytesToSend,
                        curFileBytesReceived,
                        totalBytesToSend == 0L
                    )
                }

                override fun onComplete(
                    result: TransmissionResult?,
                    r: Map<String, TransmissionResult>
                ) {
                    generalResult.set(result)
                    resultMap.set(r)
                    completed.complete(true)
                }
            })
            while (!completed.await()) {
                // Check if user cancelled
                if (isCancelled) {
                    Log.d(TAG, "User cancelled receiving file")
                    handler.cancel()
                }
            }
            if (generalResult.get() == null || generalResult.get()!!.status != org.exthmui.share.shared.misc.Constants.TransmissionStatus.COMPLETED) {
                return@coroutineScope genFailureResult(
                    (generalResult.get() as TransmissionException?)!!,
                    handler.senderInfo,
                    handler.fileInfos,
                    resultMap.get()
                )
            }
            val nsdReceiver = NsdReceiver.getInstance(it)
            val entities: MutableList<Entity?> = ArrayList(handler.fileInfos.size)
            for (fileInfo in handler.fileInfos) {
                entities.add(nsdReceiver.getEntity(fileInfo.id))
            }
            return@coroutineScope genSuccessResult(handler.senderInfo, entities)
        } ?: Result.failure()
    }

    companion object {
        const val TAG = "NsdReceivingTask"
    }
}
